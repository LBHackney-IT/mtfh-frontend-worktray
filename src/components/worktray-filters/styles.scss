@import "~lbh-frontend/lbh/base";

.worktray-filters {
  & > .govuk-details__text {
    border-left: unset;
    padding: 15px 0;
  }

  &__boxes {
    display: flex;
    flex-flow: row wrap;
    gap: 1.5em;
    justify-content: flex-start;
    align-items: flex-start;

    & > .filter-box {
      margin: 0;
    }
  }

  &__actions {
    display: flex;
    align-items: center;
    margin-top: 0;
  }

  &__actions > &__apply-button {
    margin-right: 1.5em;
  }

  &__clear-filters {
    display: flex;
    align-items: center;
    color: lbh-colour("lbh-link");
    text-decoration: underline;
  }
}

.filter-box {
  border: 2px solid lbh-colour("lbh-grey-2");
  flex: 1 0 280px;
  max-width: 380px;
  height: 100%;

  .govuk-radios__item {
    margin: 28.5px 0;
    label {
      margin: 0;
    }
  }
}

.filter-header {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  padding: 0.25em 0.5em;
  background: lbh-colour("lbh-grey-2");

  p,
  * {
    margin: 0;
  }

  .select-all {
    text-decoration: underline;
    color: lbh-colour("lbh-link");
    cursor: pointer;
  }
}

.filter-box-options {
  height: 280px;
  padding: 0 0 0 1em;
  margin: 0;
  overflow: auto;

  :last-child {
    margin-bottom: 1.5em;
  }

  .govuk-checkboxes__label {
    overflow-wrap: break-word;
  }
}
