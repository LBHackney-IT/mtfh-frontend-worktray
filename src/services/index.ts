export { default as locale } from "./locale";
export { config } from "./config";
