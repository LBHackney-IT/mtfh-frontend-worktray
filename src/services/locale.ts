const locale = {
  title: "Your tasks",
  views: {
    worktray: {
      noWorktrayResults: "No tasks found",
    },
  },
  errors: {
    unableToFetchRecord: "There was a problem retrieving the record",
    unableToFetchRecordDescription:
      "Please try again. If the issue persists, please contact support.",
    unableToFindProcess: "Invalid Process",
    unableToFindProcessDescription: "We have been unable to find a process of this type.",
    unableToFindState: "Invalid State",
    unableToFindStateDescription: "We have been unable to process a state of this type.",
  },
  components: {
    filters: {
      applyFilters: "Apply filters",
      clearFilters: "Clear filters",
      selectAll: "Select All",
      removeAll: "Remove All",
    },
    pagination: {
      previous: "Previous",
      next: "Next",
      assistiveNavigation: (page: number): string => `Page ${page}`,
      searchResultsCounter: (page: number, pageSize: number, total: number): string =>
        total === 0
          ? "Showing 0 of 0 results"
          : `Showing ${page * pageSize - pageSize + 1}â€”${Math.min(
              page * pageSize,
              total,
            )} of ${total} results`,
    },
    controls: {
      timePeriodOptions: {
        "30": "Last 30 days",
        "60": "Last 60 days",
        "3": "Last 3 months",
        "6": "Last 6 months",
        "12": "Last 12 months",
      },
      limitOption: (value: number): string => `${value} items`,
    },
  },
};

export default locale;
