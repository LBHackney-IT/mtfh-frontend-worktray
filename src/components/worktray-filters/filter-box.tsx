import React from "react";

import { Checkbox, Text } from "@mtfh/common/lib/components";

import { locale } from "../../services";

import "./styles.scss";

const { components } = locale;

export const FilterBox = ({
  filterType,
  title,
  options,
  handleSelectAll,
  handleCheckboxFilters,
  selectedFilters,
}: {
  filterType: string;
  title: string;
  options: string[];
  handleSelectAll: (value: string) => void;
  handleCheckboxFilters: (
    value: React.ChangeEvent<HTMLInputElement>,
    filterType: string,
  ) => void;
  selectedFilters: Record<string, string[]>;
}): JSX.Element => {
  return (
    <div className="filter-box">
      <div className="filter-header">
        <Text>{title}</Text>
        <button
          className="select-all lbh-link lbh-link--no-visited-state"
          onClick={() => handleSelectAll(filterType)}
        >
          {components.filters.selectAll}
        </button>
      </div>

      <div className="filter-box-options">
        {options.map((option) => {
          return (
            <Checkbox
              key={option.toLowerCase().replace(/\s/g, "-")}
              id={option.toLowerCase().replace(/\s/g, "-")}
              name={option}
              checked={selectedFilters[filterType].includes(option)}
              onChange={(event) => handleCheckboxFilters(event, filterType)}
            >
              {option}
            </Checkbox>
          );
        })}
      </div>
    </div>
  );
};
