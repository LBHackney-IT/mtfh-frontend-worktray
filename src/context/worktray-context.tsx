import React, {
  Dispatch,
  FC,
  Reducer,
  createContext,
  useContext,
  useEffect,
  useMemo,
  useReducer,
} from "react";
import { useHistory, useLocation } from "react-router-dom";

import { stringify } from "query-string";

import { AxiosSWRError, useAxiosSWR } from "@mtfh/common/lib/http";

import { config } from "../services";
import {
  LimitOptions,
  OrderByOptions,
  Process,
  ProcessSortOptions,
  TimePeriodOptions,
  WorktrayFilterOptions,
} from "../types";

interface WorktrayResults {
  results: {
    processes: Process[];
  };
  total: number;
}

export type WorktrayState = {
  page: number;
  pageSize: LimitOptions;
  timePeriod: TimePeriodOptions;
  sort: ProcessSortOptions;
  order: OrderByOptions;
  patch?: string;
  areaId?: string | null;
  processNames?: string;
  status?: string;
  results?: Process[];
  total?: number;
  error?: AxiosSWRError;
};

export type WorktrayActions =
  | {
      type: "PAGE";
      payload: number;
    }
  | {
      type: "LIMIT";
      payload: LimitOptions;
    }
  | {
      type: "TIME_PERIOD";
      payload: TimePeriodOptions;
    }
  | {
      type: "SORT";
      payload: ProcessSortOptions;
    }
  | {
      type: "FILTER";
      payload: {
        type: WorktrayFilterOptions;
        payload: string;
      };
    }
  | {
      type: "ORDER";
      payload: OrderByOptions;
    };

export interface WorktrayProviderProps {
  initial: Partial<WorktrayState>;
}

const getInitialState = (
  options: WorktrayProviderProps["initial"] = {},
): WorktrayState => {
  return {
    page: options.page || 1,
    pageSize: options.pageSize || LimitOptions.SMALL,
    timePeriod: options.timePeriod || TimePeriodOptions.DAYS_30,
    sort: options.sort || ProcessSortOptions.STATUS,
    order: options.order || OrderByOptions.ASC,
    processNames: options.processNames || undefined,
    patch: options.patch || undefined,
    areaId: options.areaId || undefined,
    status: options.status || undefined,
  };
};

export const WorktrayContext = createContext<{
  state: WorktrayState;
  dispatch: Dispatch<WorktrayActions>;
}>({
  state: getInitialState(),
  dispatch: () => {},
});

export const WorktrayProvider: FC<WorktrayProviderProps> = ({ children, initial }) => {
  const reducer: Reducer<WorktrayState, WorktrayActions> = (state, action) => {
    switch (action.type) {
      case "PAGE": {
        return {
          ...state,
          page: action.payload,
        };
      }
      case "LIMIT": {
        if (Object.values(LimitOptions).includes(action.payload)) {
          return {
            ...state,
            pageSize: action.payload,
          };
        }
        return state;
      }
      case "TIME_PERIOD": {
        if (Object.values(TimePeriodOptions).includes(Number(action.payload))) {
          return {
            ...state,
            timePeriod: action.payload,
          };
        }
        return state;
      }
      case "SORT": {
        if (Object.values(ProcessSortOptions).includes(action.payload)) {
          return {
            ...state,
            sort: action.payload,
          };
        }
        return state;
      }
      case "ORDER": {
        if (Object.values(OrderByOptions).includes(action.payload)) {
          return {
            ...state,
            order: action.payload,
          };
        }
        return state;
      }
      case "FILTER": {
        return {
          ...state,
          [action.payload.type]: action.payload.payload,
        };
      }
      default: {
        return state;
      }
    }
  };

  const initialState = getInitialState(initial);
  const [state, dispatch] = useReducer(reducer, initialState);
  const { ...query } = state;

  const { data, error } = useAxiosSWR<WorktrayResults>(
    `${config.searchApiUrl}/search/processes?${stringify(
      {
        searchText: query.patch,
        page: query.page,
        pageSize: query.pageSize,
        timePeriod: query.timePeriod,
        processNames: query.processNames,
        status: query.status,
        sortBy: query.sort,
        isDesc: query.order === "asc",
      },
      {
        skipEmptyString: true,
      },
    )}`,
    {
      refreshWhenHidden: false,
      revalidateOnFocus: false,
    },
  );

  useEffect(() => {
    if (data?.total && data.total > 0) {
      const totalPages = Math.ceil(data.total / state.pageSize);
      if (state.page > totalPages) {
        dispatch({ type: "PAGE", payload: totalPages });
      }
    }
  }, [state.page, data?.total, state.pageSize]);

  const value = useMemo(
    () => ({
      state: {
        ...state,
        results: data?.results.processes,
        total: data?.total,
        error,
      },
      dispatch,
    }),
    [state, data, error, dispatch],
  );

  return <WorktrayContext.Provider value={value}>{children}</WorktrayContext.Provider>;
};

export const PersistWorktrayContextURL = ({
  sessionKey,
}: {
  sessionKey?: string;
}): null => {
  const {
    state: { page, pageSize, order, sort, timePeriod, patch, processNames, status },
  } = useContext(WorktrayContext);
  const { push } = useHistory();

  useEffect(() => {
    const query = new URLSearchParams();
    if (page > 1) {
      query.set("p", String(page));
    }

    if (pageSize !== LimitOptions.SMALL) {
      query.set("l", String(pageSize));
    }

    query.set("t", String(timePeriod));

    if (order && order !== OrderByOptions.ASC) {
      query.set("o", order);
    }

    query.set("sort", String(sort));

    if (patch) {
      query.set("patch", patch);
    }

    if (processNames) {
      query.set("processNames", processNames);
    }

    if (status) {
      query.set("status", status);
    }

    const path = `?${query.toString()}`;
    push(path);

    if (sessionKey) {
      window.sessionStorage.setItem(sessionKey, path);
    }
  }, [
    page,
    pageSize,
    order,
    sort,
    sessionKey,
    push,
    timePeriod,
    patch,
    processNames,
    status,
  ]);

  return null;
};

export const WorktrayURLProvider: FC<{
  sessionKey?: string;
  patchId?: string | null;
  areaId?: string | null;
}> = ({ children, sessionKey, patchId, areaId }): JSX.Element => {
  const { search } = useLocation();
  const query = new URLSearchParams(search);
  const page = Number(query.get("p") || 1);
  const pageSize = Number(query.get("l") || 10);
  const timePeriod = (query.get("t") || "") as TimePeriodOptions;
  const patch = query.get("patch") || patchId || "";
  const processNames = query.get("processNames") || "";
  const status = query.get("status") || "";
  const sortBy = query.get("sort");
  const orderBy = query.get("o");

  const sort = useMemo(() => {
    if (sortBy === "name") {
      return ProcessSortOptions.NAME;
    }
    return ProcessSortOptions.STATUS;
  }, [sortBy]);

  const order = useMemo(() => {
    if (orderBy === "desc") {
      return OrderByOptions.DESC;
    }
    return OrderByOptions.ASC;
  }, [orderBy]);

  return (
    <WorktrayProvider
      initial={{
        page,
        sort,
        order,
        pageSize,
        timePeriod,
        patch,
        areaId,
        processNames,
        status,
      }}
    >
      {children}
      <PersistWorktrayContextURL sessionKey={sessionKey} />
    </WorktrayProvider>
  );
};
