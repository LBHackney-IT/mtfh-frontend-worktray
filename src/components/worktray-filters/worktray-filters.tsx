import React, { useCallback, useContext, useState } from "react";

import { Button, Details } from "@mtfh/common/lib/components";
import { useAxiosSWR } from "@mtfh/common/lib/http";
import { Process, Status, processes } from "@mtfh/processes";

import { WorktrayContext } from "../../context/worktray-context";
import { config } from "../../services";
import locale from "../../services/locale";
import { WorktrayFilterOptions } from "../../types";
import { Patch } from "../../types/patch";
import { FilterBox, Option } from "./filter-box";

import "./styles.scss";

const { components } = locale;

export const WorktrayFilters = (): JSX.Element => {
  const {
    dispatch,
    state: { patch, patchId, process, status },
  } = useContext(WorktrayContext);

  const { data: patchData } = useAxiosSWR<Patch>(
    `${config.patchesAndAreasApiUrl}/patch/${patchId}`,
  );

  const { data: patches } = useAxiosSWR<Patch[]>(
    `${config.patchesAndAreasApiUrl}/patch?parentId=${patchData?.parentId}`,
  );

  const filters: { type: string; title: string; options: Option[] }[] = [
    {
      type: "process",
      title: "Processes",
      options: (Object.values(Process) as string[]).map((processName) => {
        const value = processes[processName].name;
        return {
          key: value.toLowerCase().replace(/\s/g, "-"),
          value,
        };
      }),
    },
  ];

  if (patches) {
    filters.push({
      type: "patch",
      title: "Patches",
      options: patches?.map((item) => ({ key: item.id, value: item.name })),
    });
  }

  filters.push({
    type: "status",
    title: "Process Status",
    options: (Object.values(Status) as string[]).map((value) => {
      return {
        key: value.toLowerCase().replace(/\s/g, "-"),
        value,
      };
    }),
  });

  const [selectedFilters, setSelectedFilters] = useState<
    Record<WorktrayFilterOptions, string[]>
  >({
    patch: patch?.split(",") || [],
    process: process?.split(",") || [],
    status: status?.split(",") || [],
  });

  const handleCheckboxFilters = (
    event: React.ChangeEvent<HTMLInputElement>,
    filterType: string,
  ) => {
    const updatedFilters: string[] = [...selectedFilters[filterType]];
    if (event.target.checked) {
      updatedFilters.push(event.target.name);
    } else {
      updatedFilters.splice(selectedFilters[filterType].indexOf(event.target.name), 1);
    }
    setSelectedFilters({
      ...selectedFilters,
      [filterType]: updatedFilters,
    });
  };

  const handleSelectAll = (filterType: string) => {
    setSelectedFilters({
      ...selectedFilters,
      [filterType]:
        filters
          .find((filter) => filter.type === filterType)
          ?.options.map((option) => option.key) || [],
    });
  };

  const applyFilters = useCallback(
    (data) => {
      (Object.keys(data) as WorktrayFilterOptions[]).forEach((filterType) => {
        dispatch({
          type: "FILTER",
          payload: {
            type: filterType,
            payload: String(data[filterType]),
          },
        });
      });
    },
    [dispatch],
  );

  const clearFilters = () => {
    Object.keys(selectedFilters).forEach((filterType) => {
      selectedFilters[filterType] = [];
    });
    applyFilters(selectedFilters);
  };

  return (
    <Details className="worktray-filters" title="Filter by" open>
      <>
        <div className="worktray-filters__boxes">
          {filters.map(({ title, options, type }) => {
            return (
              <FilterBox
                key={title}
                filterType={type}
                title={title}
                options={options}
                handleSelectAll={handleSelectAll}
                handleCheckboxFilters={handleCheckboxFilters}
                selectedFilters={selectedFilters}
              />
            );
          })}
        </div>

        <div className="worktray-filters__actions">
          <Button
            className="worktray-filters__apply-button"
            onClick={() => applyFilters(selectedFilters)}
          >
            {components.filters.applyFilters}
          </Button>
          <button
            className="worktray-filters__clear-filters select-all lbh-link lbh-link--no-visited-state"
            onClick={() => clearFilters()}
          >
            {components.filters.clearFilters}
          </button>
        </div>
      </>
    </Details>
  );
};
