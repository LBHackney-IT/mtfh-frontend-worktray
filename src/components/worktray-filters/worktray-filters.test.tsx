import React from "react";

import { render } from "@hackney/mtfh-test-utils";
import { fireEvent, screen } from "@testing-library/react";

import { WorktrayURLProvider } from "../../context/worktray-context";
import { locale } from "../../services";
import { WorktrayFilters } from "./worktray-filters";

const { components } = locale;

const filters = [
  {
    type: "processes",
    title: "Processes",
    options: ["Process 1", "Process 2", "Process 3"],
  },
  {
    type: "patches",
    title: "Patches",
    options: ["CP1", "CP2", "CP3"],
  },
];

describe("worktray-filters", () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  test("it renders WorktrayFilters correctly", async () => {
    const { container } = render(<WorktrayFilters />);
    expect(container).toMatchSnapshot();
  });

  test("it selects and unselects filter options", async () => {
    render(
      <WorktrayURLProvider sessionKey="test">
        <WorktrayFilters />
      </WorktrayURLProvider>,
      {
        url: "/?t=30&sort=status&process=%2CProcess+2&patch=%2CCP1",
        path: "/",
      },
    );
    await expect(
      screen.findByLabelText(filters[0].options[0]),
    ).resolves.not.toBeChecked();
    fireEvent.click(screen.getByLabelText(filters[0].options[0]));
    await expect(screen.findByLabelText(filters[0].options[0])).resolves.toBeChecked();
    fireEvent.click(screen.getByText(locale.components.filters.applyFilters));
    expect(window.sessionStorage.getItem("test")).toBe(
      "?t=30&sort=status&patch=%2CCP1&process=%2CProcess+2%2CProcess+1",
    );
    fireEvent.click(screen.getByLabelText(filters[0].options[0]));
    await expect(
      screen.findByLabelText(filters[0].options[0]),
    ).resolves.not.toBeChecked();
  });

  test("it selects all options and clears filters correctly", async () => {
    render(
      <WorktrayURLProvider sessionKey="test">
        <WorktrayFilters />
      </WorktrayURLProvider>,
    );

    await expect(
      screen.findByLabelText(filters[0].options[0]),
    ).resolves.not.toBeChecked();
    await expect(
      screen.findByLabelText(filters[0].options[1]),
    ).resolves.not.toBeChecked();
    await expect(
      screen.findByLabelText(filters[0].options[2]),
    ).resolves.not.toBeChecked();

    fireEvent.click(screen.getAllByText(components.filters.selectAll)[0]);

    await expect(screen.findByLabelText(filters[0].options[0])).resolves.toBeChecked();
    await expect(screen.findByLabelText(filters[0].options[1])).resolves.toBeChecked();
    await expect(screen.findByLabelText(filters[0].options[2])).resolves.toBeChecked();

    fireEvent.click(screen.getByText(components.filters.clearFilters));

    await expect(
      screen.findByLabelText(filters[0].options[0]),
    ).resolves.not.toBeChecked();
    await expect(
      screen.findByLabelText(filters[0].options[1]),
    ).resolves.not.toBeChecked();
    await expect(
      screen.findByLabelText(filters[0].options[2]),
    ).resolves.not.toBeChecked();
  });
});
