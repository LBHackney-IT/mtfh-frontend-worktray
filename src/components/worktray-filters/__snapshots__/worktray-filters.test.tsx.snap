// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`worktray-filters it renders WorktrayFilters correctly with patches 1`] = `
<div>
  <details
    class="govuk-details lbh-details worktray-filters"
    data-module="govuk-details"
    data-testid="mtfh-details"
    id="mtfh-details"
  >
    <summary
      class="govuk-details__summary"
    >
      <span
        class="govuk-details__summary-text"
      >
        Filter by
      </span>
    </summary>
    <div
      class="govuk-details__text"
    >
      <div
        class="worktray-filters__boxes"
      >
        <div
          class="filter-box"
        >
          <div
            class="filter-header"
          >
            <p
              class="lbh-body-m"
            >
              Processes
            </p>
            <button
              class="select-all lbh-link lbh-link--no-visited-state"
              data-testid="filter-box-select-all"
            >
              Select All
            </button>
          </div>
          <div
            class="filter-box-options"
          >
            <div
              class="govuk-checkboxes__item"
            >
              <input
                class="govuk-checkboxes__input"
                data-testid="filter-option-soletojoint"
                id="soletojoint"
                name="soletojoint"
                type="checkbox"
              />
              <label
                class="govuk-label govuk-checkboxes__label"
                for="soletojoint"
              >
                Sole to Joint
              </label>
            </div>
            <div
              class="govuk-checkboxes__item"
            >
              <input
                class="govuk-checkboxes__input"
                data-testid="filter-option-changeofname"
                id="changeofname"
                name="changeofname"
                type="checkbox"
              />
              <label
                class="govuk-label govuk-checkboxes__label"
                for="changeofname"
              >
                Change of Name
              </label>
            </div>
          </div>
        </div>
        <div
          class="filter-box"
        >
          <div
            class="filter-header"
          >
            <p
              class="lbh-body-m"
            >
              Patches
            </p>
          </div>
          <div
            class="filter-box-options"
          >
            <div
              class="govuk-radios lbh-radios"
            >
              <div
                class="govuk-radios__item"
              >
                <input
                  class="govuk-radios__input"
                  id="show-all"
                  type="radio"
                  value="show-all"
                />
                <label
                  class="govuk-label govuk-radios__label"
                  for="show-all"
                >
                  Show All
                </label>
              </div>
              <div
                class="govuk-radios__item"
              >
                <input
                  class="govuk-radios__input"
                  id="19400185-a8d2-4f3f-b217-dc5d485a1211"
                  type="radio"
                  value="19400185-a8d2-4f3f-b217-dc5d485a1211"
                />
                <label
                  class="govuk-label govuk-radios__label"
                  for="19400185-a8d2-4f3f-b217-dc5d485a1211"
                >
                  AB1
                </label>
              </div>
              <div
                class="govuk-radios__item"
              >
                <input
                  class="govuk-radios__input"
                  id="19400185-a8d2-4f3f-b217-dc5d485a1212"
                  type="radio"
                  value="19400185-a8d2-4f3f-b217-dc5d485a1212"
                />
                <label
                  class="govuk-label govuk-radios__label"
                  for="19400185-a8d2-4f3f-b217-dc5d485a1212"
                >
                  CP6
                </label>
              </div>
              <div
                class="govuk-radios__item"
              >
                <input
                  class="govuk-radios__input"
                  id="19400185-a8d2-4f3f-b217-dc5d485a1210"
                  type="radio"
                  value="19400185-a8d2-4f3f-b217-dc5d485a1210"
                />
                <label
                  class="govuk-label govuk-radios__label"
                  for="19400185-a8d2-4f3f-b217-dc5d485a1210"
                >
                  CP7
                </label>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="worktray-filters__actions"
      >
        <button
          class="govuk-button lbh-button worktray-filters__apply-button"
          type="button"
        >
          Apply filters
        </button>
        <button
          class="worktray-filters__clear-filters select-all lbh-link lbh-link--no-visited-state"
        >
          Clear filters
        </button>
      </div>
    </div>
  </details>
</div>
`;
